flink-tests/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterStopWithSavepointITCase.java

```diff
@@ -242,6 +258,27 @@ public void testRestartCheckpointCoordinatorIfStopWithSavepointFails() throws Ex
         assertTrue(checkpointsToWaitFor.await(60L, TimeUnit.SECONDS));
     }
 
+    private void stopWithSavepointAndWait(boolean terminate)
+            throws InterruptedException, ExecutionException {
+        final int attempts = 15;
+        for (int i = 0; i < attempts; i++) {
+            try {
+                stopWithSavepoint(terminate).get();
+                return;
+            } catch (Exception ex) {
+                Optional<CheckpointException> checkpointExceptionOptional =
+                        ExceptionUtils.findThrowable(ex, CheckpointException.class);
+                if (!checkpointExceptionOptional.isPresent()
+                        || checkpointExceptionOptional.get().getCheckpointFailureReason()
+                                != NOT_ALL_REQUIRED_TASKS_RUNNING) {
+                    throw ex;
+                }
+                // back off before triggering savepoint again
+                Thread.sleep(100);
+            }
+        }
+    }
+
```

Comment:
When do we need so many tries? Why wouldn't it work if we know that everything is running on the first try?

Reply:
It would work if we knew everything is running, however I don't know how to achieve that. I could not find a way to reliably get that information.

The number of tries is arbitrary. There is no true logic behind picking 15, I can decrease the number.

There is one additional trick that actually we expect the `stopWithSavepoint` to fail in some of the tests. However we never expect it to fail with the `NOT_ALL_REQUIRED_TASKS_RUNNING` and we only ever retry if that is the cause.